torch/overrides.py

```diff
@@ -1851,31 +1818,69 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
         raise NotImplementedError()
 
     def __enter__(self):
-        old = _get_torch_function_mode()
-        if hasattr(self, "inner"):
-            raise RuntimeError(f"{self} has already been used as a mode. Please use a fresh version or use restore")
-        else:
-            self.inner = old
-            if old is None:
-                self.ancestors = set()
-            else:
-                self.ancestors = self.inner.ancestors.union({self.inner})
-        _set_torch_function_mode(self)
+        _push_mode(self)
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        _set_torch_function_mode(self.inner)
-
-    @contextlib.contextmanager
-    def restore(self):
-        return _restore_mode(self, mode_info=_TorchFunctionModeInfo())
+        _pop_mode()
 
     @classmethod
     def push(cls, *args, **kwargs):
         warnings.warn("`Mode.push()` is no longer necessary and can be replaced with just `with Mode()`")
         instance = cls(*args, **kwargs)
         return instance
 
+
+# WARNING: If you access this list to see all the active modes in order, do not update the list in place
+# since it alters the actual mode stack
+threadLocal = threading.local()
```

Comment:
One final change was to make each of the mode stacks thread safe. I was seeing multiprocessing tests pass without this change (?) but figured it's better to be safe here

Reply:
It looks like the multiprocessing tests can pass if all the threads in the pool run sequentially. Maybe that's just what happened. I don't know the python multithreading API well, but to test this deterministically, we could have two threads, A and B, that communicate with each other. A sets up a mode, and then with the mode active, communicates to B and blocks until it hears a response. B then records the number of modes active, and then communicates back to A.

Reply:
Discussed offline: Python multiprocessing is always multi-process (to avoid the GIL), so testing thread-localness doesn't actually work there. We would either need to write a test using the python threading module or just believe in our code

Reply:
Pulled test because of this. Tested locally with threading library.

Also so I remember, from separate offline conversation, could have also chose to save this as a PyObject on the TLS and update that